tensorflow-training/src/main/java/org/tensorflow/training/optimizers/Optimizer.java

```diff
@@ -0,0 +1,279 @@
+  }
+
+  public Op minimize(Operand<?> loss, String name) {
+    List<GradAndVar<?>> gradsAndVars = computeGradients(loss);
+
+    return applyGradients(gradsAndVars, name);
+  }
+
+  public <T extends TType> List<GradAndVar<?>> computeGradients(Operand<?> loss) {
+    List<Operation> variables = new ArrayList<>();
+    Iterator<Operation> opItr = graph.operations();
+    while (opItr.hasNext()) {
+      Operation op = opItr.next();
+      if (op.type().equals(VARIABLE_V2)) {
+        variables.add(op);
+      }
+    }
+
+    Output<?>[] variableOutputArray = new Output[variables.size()];
+    for (int i = 0; i < variables.size(); i++) {
+      // First output of a variable is it's output.
+      variableOutputArray[i] = variables.get(i).output(0);
+    }
+
+    Output<?>[] gradients = graph.addGradients(loss.asOutput(), variableOutputArray);
+    List<GradAndVar<? extends TType>> gradVarPairs = new ArrayList<>();
+
+    for (int i = 0; i < variableOutputArray.length; i++) {
+      @SuppressWarnings("unchecked")
+      Output<T> typedGrad = (Output<T>) gradients[i];
+      @SuppressWarnings("unchecked")
+      Output<T> typedVar = (Output<T>) variableOutputArray[i];
+      gradVarPairs.add(new GradAndVar<>(typedGrad, typedVar));
+    }
+
+    return gradVarPairs;
+  }
+
+  public Op applyGradients(List<GradAndVar<? extends TType>> gradsAndVars, String name) {
+    List<Output<? extends TType>> variables = gradsAndVars.stream().map(GradAndVar::getVariable)
+        .collect(Collectors.toList());
+
+    createSlots(variables);
+
+    Optional<Operand<? extends TType>> prepOp = prepare(name + "/prepare");
+
+    List<Operand<? extends TType>> updateOps = new ArrayList<>();
+    prepOp.ifPresent(updateOps::add);
+    for (GradAndVar<? extends TType> pair : gradsAndVars) {
+      updateOps.add(applyDense(pair));
+    }
+
+    return finish(updateOps, name);
+  }
+
+  /**
+   * Gets the slot associated with the specified variable and slot name.
+   *
+   * @param var      The variable to lookup.
+   * @param slotName The slot name.
+   * @return The slot or {@link Optional#empty}.
+   */
+  public <T extends TType> Optional<Variable<T>> getSlot(Output<T> var, String slotName) {
+    return getSlot(var.op().name(), slotName);
+  }
+
+  /**
+   * Gets the slot associated with the specified variable and slot name.
+   *
+   * @param varName  The variable to lookup.
+   * @param slotName The slot name.
+   * @return The slot or {@link Optional#empty}.
+   */
+  private <T extends TType> Optional<Variable<T>> getSlot(String varName, String slotName) {
+    Map<String, Variable<? extends TType>> variables = slots.get(slotName);
+    if (variables != null) {
+      Variable<? extends TType> slot = variables.get(varName);
+      if (slot != null) {
+        @SuppressWarnings("unchecked") // This method should only be called when the type is known.
+            Optional<Variable<T>> opt = Optional.of((Variable<T>) slot);
```

Comment:
Nit: either it is GitHub or there is a wrong left margin to this line. Also, no need of else blocks below since you return in previous blocks.

Reply:
The Google code formatter in IntelliJ keeps doing this, I'll keep moving it back. It's because it treats the annotation and the variable creation as the same statement, and thinks it's a line continuation.